trajoptlib/src/DifferentialTrajectoryGenerator.cpp

```diff
@@ -0,0 +1,273 @@
+            constraint);
+      }
+    }
+  }
+
+  // Minimize total time
+  sleipnir::Variable T_tot = 0;
+  for (size_t sgmtIndex = 0; sgmtIndex < N.size(); ++sgmtIndex) {
+    auto& dt_sgmt = dt.at(sgmtIndex);
+    auto N_sgmt = N.at(sgmtIndex);
+    auto T_sgmt = dt_sgmt * static_cast<int>(N_sgmt);
+    T_tot += T_sgmt;
+
+    problem.SubjectTo(dt_sgmt >= 0);
+    dt_sgmt.SetValue(5.0 / N_sgmt);
+  }
+  problem.Minimize(std::move(T_tot));
+
+  // Apply kinematics constraints
+  for (size_t wptIndex = 1; wptIndex < wptCnt; ++wptIndex) {
+    size_t N_sgmt = N.at(wptIndex - 1);
+    auto dt_sgmt = dt.at(wptIndex - 1);
+
+    for (size_t sampIndex = 0; sampIndex < N_sgmt; ++sampIndex) {
+      size_t index = GetIndex(N, wptIndex, sampIndex);
+
+      Translation2v x_n{x.at(index), y.at(index)};
+      Translation2v x_n_1{x.at(index - 1), y.at(index - 1)};
+
+      Rotation2v theta_n{thetacos.at(index), thetasin.at(index)};
+      Rotation2v theta_n_1{thetacos.at(index - 1), thetasin.at(index - 1)};
+
+      Translation2v v_n{(vL.at(index) + vR.at(index)) / 2, 0.0};
+      Translation2v v_n_1{(vL.at(index - 1) + vR.at(index - 1)) / 2, 0.0};
+
+      auto omega_n = (vR.at(index) - vL.at(index)) / path.drivetrain.trackwidth;
+      auto omega_n_1 =
+          (vR.at(index - 1) - vL.at(index - 1)) / path.drivetrain.trackwidth;
+
+      Translation2v a_n{((vL.at(index) + vR.at(index)) / 2 -
+                         (vL.at(index - 1) + vR.at(index - 1)) / 2) /
+                            dt.at(index),
+                        0.0};
+      auto alpha_n =
+          (vR.at(index) - vL.at(index)) / path.drivetrain.trackwidth -
+          (vR.at(index - 1) - vL.at(index - 1)) / path.drivetrain.trackwidth /
+              dt.at(index);
+
+      problem.SubjectTo(x_n_1 + v_n * dt_sgmt == x_n);
+      problem.SubjectTo((theta_n - theta_n_1) == Rotation2v{omega_n * dt_sgmt});
+      problem.SubjectTo(v_n_1 + a_n * dt_sgmt == v_n);
+      problem.SubjectTo(omega_n_1 + alpha_n * dt_sgmt == omega_n);
+    }
+  }
+
+  for (size_t index = 0; index < sampTot; ++index) {
+    // Apply dynamics constraints
+    problem.SubjectTo(tauL.at(index) == path.drivetrain.moi * vL.at(index));
+    problem.SubjectTo(tauR.at(index) == path.drivetrain.moi * vR.at(index));
+  }
+
+  for (size_t wptIndex = 0; wptIndex < wptCnt; ++wptIndex) {
+    for (auto& constraint : path.waypoints.at(wptIndex).waypointConstraints) {
+      // First index of next wpt - 1
+      size_t index = GetIndex(N, wptIndex + 1, 0) - 1;
+      Pose2v pose{
+          x.at(index), y.at(index), {thetacos.at(index), thetasin.at(index)}};
+      Translation2v linearVelocity{(vL.at(index) + vR.at(index)) / 2, 0.0};
+      auto angularVelocity =
+          (vR.at(index) - vL.at(index)) / path.drivetrain.trackwidth;
+      Translation2v linearAcceleration{
+          ((vL.at(index) + vR.at(index)) / 2 -
+           (((index > 0) ? vL.at(index - 1) : 0.0) +
+            ((index > 0) ? vR.at(index - 1) : 0.0)) /
```

Comment:
An if statement would be easier to read. Also, you shouldn't need to use finite backward difference here because linear and angular acceleration should be decision variables (at least, they were for swerve).